### DISCUSSION TITLE
Is there a way to have Rewrites or Proxy?

### DISCUSSION THREAD

**Discussion by andrewdavid48** [TanStack/router#2399] [Q&A] [answered]
This might be a dumb question, but I couldn't find any docs for it, so I might as well ask.

My team is moving away from nextjs (pages) and moving to TSR (tanstack router with start/vinxi).
One of the difficulties we're seeing is that we couldn't find any docs on rewrites/proxies.

The reason this is a blocker for us is that we need to have the API set cookies from the same domain as the UI (due to multiple envs, legacy auth issue and others).

FWIW, i tried the vite server proxy, but it seems that tanstack router intercepts the request before vite proxy can (assumption).

In our `nextjs.config.ts` we have the following:
```ts
// nextjs.config.ts
{
  rewrites() {
    return [
      { source: "/api/avatar/:path*", destination: `${API_URL}/members/:path*/avatar` },
      { source: "/api/:path*", destination: `${API_URL}/:path*` },
    ];
  },
}
```

We played around a bit and have a very hacky solution which is very obnoxious. We do send headers and other through, but not in this example.
Hopefully there's a better way that this:
```ts
// routes/api/$.ts
// ...

function constructUrl(splat: any) {
  return `${getApiBaseUrl()}/${
    Array.isArray(splat) ? splat.join("/") : typeof splat === "object" ? Object.values(splat).join("/") : splat
  }`;
}

const createResponse = (data: any, headers: Headers, status: number) => 
      new Response(JSON.stringify(data), { headers, status });

export const Route = createAPIFileRoute("/api/$")({
  GET: async ({ params }) => {
    const { _splat } = params;
    const url = constructUrl(_splat);
    const res = await api.get(url);

    return createResponse(res.data, res.headers, res.status);
  },
  POST: async ({ params, request }) => {
    const { _splat } = params;
    const body = await request.json();
    const url = constructUrl(_splat);
    const res = await api.post(url, body);

    return createResponse(res.data, res.headers, res.status);
  },
});
```

**Comment by schiller-manuel** [comment, marked as answer]
as of v1.58.7, we have the following:

TanStack Start now exposes the `routeRules` config option from nitro: https://nitro.unjs.io/guide/routing#route-rules

example:

```tsx
import { defineConfig } from '@tanstack/start/config'

export default defineConfig({
  server: {
    routeRules: {
      "/proxy/example": { proxy: { to: "https://example.com"} },
    }
  },
})

**Reply by andrewdavid48** [reply]
Thank you!

**Reply by rijk** [reply]
Hey @schiller-manuel , I tested and this doesn't seem to support rewriting parameters. Example:

```ts

export default defineConfig({
  server: {
    routeRules: {
      '/test/**': { redirect: { to: '/**', statusCode: 302 } },
    },
  },
})
```

`GET localhost:3004/test/foo/bar `

Expected: `localhost:3004/foo/bar`
Actual: `localhost:3004/**`

**Reply by schiller-manuel** [reply]
@rijk seems to be a different use case. can you please create a new issue/discussion for what you want to do?

**Reply by rijk** [reply]
Twofold:

1. Proxy internally from `/[slug]/[id]` to `/[id]` (`[slug]` changes when renaming the item and is not used for querying) — I guess this can also be solved by ignoring the $slug parameter
2. Proxy externally as a [reverse proxy](https://posthog.com/docs/advanced/proxy/nextjs); `/analytics/**` → `https://eu.posthog.com/**`

**Reply by ssijak** [reply]
@rijk dont use analytics in the url, blockers could block the request, use some random word that has nothing to do with analytics

**Reply by carderne** [reply]
@rijk for the Posthog use-case: why are you trying to use `redirect` instead of `proxy`? The proxy example in the answer above works fine for Posthog.

**Reply by Emilios1995** [reply]
@schiller-manuel How should we handle this now that `'@tanstack/start/config'` is not used anymore?

**Reply by JeffersonCarvalh0** [reply]
@Emilios1995 I don't know if you've already found this, but there's a `proxy` key from the usual vite's `defineConfig` which can be used:

```ts
import { defineConfig } from 'vite'

export default defineConfig({
  server: {
    proxy: {
      "/api": {
        target: process.env.VITE_PUBLIC_API_URL,
        changeOrigin: true,
        secure: false,
        rewrite: (path) => path.replace(/^\/api/, ""),
      },
    },
  },
})
```

More info here: https://vite.dev/config/server-options.html#server-proxy

**Reply by devrayat000** [reply]
@schiller-manuel How do we implement sub-domain rewrites?

**Reply by schiller-manuel** [reply]
@devrayat000 can you please share more details of what you want to do?

**Reply by devrayat000** [reply]
@schiller-manuel I want the following for example
- domain.com -> /
- admin.domain.com/* -> /admin/*
- auth.domain.com/* -> /auth/*

**Reply by luksch42** [reply]
@devrayat000 did you solve it?

**Reply by nahtnam** [reply]
@JeffersonCarvalh0 is the vite config option working for you? Works for me in development but does not work in production (just returns 404)

**Reply by luksch42** [reply]
Thats only for dev mode @nahtnam

**Reply by nahtnam** [reply]
Gotcha,

In case anyone else is looking for this, I spent a while trying different libraries but this ended up working the best:
```
import { createServerFileRoute } from "@tanstack/react-start/server";
import { Hono } from "hono";
import { proxy } from "hono/proxy";
import { handle } from "hono/vercel";

const app = new Hono().basePath("/docs");

app.all("*", (ctx) => {
  const url = ctx.req.url.replace(
    ctx.req.header("Host")!,
    "INSERT_URL_HERE",
  );
  return proxy(url);
});

async function handler({ request }: { request: Request }) {
  return handle(app)(request);
}

export const methods = {
  DELETE: handler,
  GET: handler,
  HEAD: handler,
  PATCH: handler,
  POST: handler,
  PUT: handler,
};

export const ServerRoute = createServerFileRoute("/docs/$").methods(methods);

```

**Reply by JeffersonCarvalh0** [reply]
> @JeffersonCarvalh0 is the vite config option working for you? Works for me in development but does not work in production (just returns 404)

Unfortunately no, it doesn't work for me in production either. I hadn't tried when I wrote that answer.
I ended up ditching reverse proxy completely and enabling CORS in my backend because of that =/

**Reply by devrayat000** [reply]
> @devrayat000 did you solve it?

No. The developer just created a new framework, created some hype, and then disappeared when they were asked valid questions. There are a small number of threads on this issue, and every time the developers of the framework have ghosted everyone.

**Reply by andrewdavid48** [reply]
> > @devrayat000 did you solve it?
> 
> No. The developer just created a new framework, created some hype, and then disappeared when they were asked valid questions. There are a small number of threads on this issue, and every time the developers of the framework have ghosted everyone.

I disagree with you that the devs created hype and disappeared. The tanstack community is very active on discord and devs have worked tirelessly on major changes (see devinxi). I encourage you to join the discord community. It's possible your question has already been answered there.

**Comment by xuxucode** [comment]
How to rewrite `/@username` to `/users/username`? `username` is a dynamic url parameter.

In nextjs, we are using below configuration:
```ts
// next.config.ts
const nextConfig: NextConfig = {
  async rewrites() {
    return [
      {
        source: "/@:username/:path*",
        destination: "/users/:username/:path*",
      },
    ]
  },
}
```

**Reply by CanRau** [reply]
With the new version currently in alpha you can have a file route called `/@{username}` and possibly wouldn't need the rewrite for this case anymore https://github.com/TanStack/router/issues/707#issuecomment-2873811856

If you still want to group you could use a [pathless layout route](https://tanstack.com/router/latest/docs/framework/react/routing/routing-concepts#pathless-layout-routes)

**Comment by bayasdev** [comment]
This works for me

```typescript
import {
	createServerFileRoute,
	getRequestIP,
	proxyRequest,
} from "@tanstack/react-start/server";

const BACKEND_URL = import.meta.env.VITE_SERVER_URL || "http://localhost:8000";

export const ServerRoute = createServerFileRoute("/api/$").methods((api) => {
	const handler = api.handler(async ({ request, params }) => {
		const url = new URL(request.url);
		const targetUrl = `${BACKEND_URL}/api/${params._splat ?? ""}${url.search}`;

		const ip = getRequestIP({ xForwardedFor: true });

		await proxyRequest(targetUrl, { headers: { "X-Forwarded-For": ip } });

		return new Response();
	});

	return {
		GET: handler,
		POST: handler,
		PUT: handler,
		PATCH: handler,
		DELETE: handler,
		OPTIONS: handler,
		HEAD: handler,
	};
});
```

**Reply by luixo** [reply]
This was removed in v1.132.0, please support it's return: https://github.com/TanStack/router/issues/5359

**Comment by sunxyw** [comment]
This is what I am doing right now, using the API route. Sharing it in case anyone can get some inspiration.

```ts
import { createFileRoute } from "@tanstack/react-router";

const API_HOST = "us.i.posthog.com" as const;
const ASSET_HOST = "us-assets.i.posthog.com" as const;

const posthogProxy = async ({ request }: { request: Request }) => {
  const url = new URL(request.url);
  const hostname = url.pathname.startsWith("/api/ingest/static/")
    ? ASSET_HOST
    : API_HOST;

  const newUrl = new URL(url);
  newUrl.protocol = "https";
  newUrl.hostname = hostname;
  newUrl.port = "443";
  newUrl.pathname = newUrl.pathname.replace(/^\/api\/ingest/, "");

  const headers = new Headers(request.headers);
  headers.set("host", hostname);

  const response = await fetch(newUrl, {
    method: request.method,
    headers,
    body: request.body,
  });

  return new Response(response.body, {
    status: response.status,
    statusText: response.statusText,
    headers: response.headers,
  });
};

export const Route = createFileRoute("/api/ingest/$")({
  server: {
    handlers: {
      ALL: posthogProxy,
      GET: posthogProxy,
      POST: posthogProxy,
      PUT: posthogProxy,
      PATCH: posthogProxy,
      DELETE: posthogProxy,
      OPTIONS: posthogProxy,
      HEAD: posthogProxy,
    },
  },
});
```

**Comment by uncvrd** [comment]
Sharing my slightly updated solution to @sunxyw his solution didn't work for some reason due to the fact that 

1. Passing a body to a fetch required setting the duplex param 🤷 
2. Passing compressed data without removing the content encoding headers causes a 200 content encoding error on the frontend

```
import { POSTHOG_API_HOST_PREFIX, POSTHOG_ASSETS_HREF, POSTHOG_PUBLIC_HREF } from "@foundrydev/api-clients/shared/posthog/constants.js"
import { createFileRoute } from "@tanstack/react-router"

const handler = async ({ request }: { request: Request }) => {
    const url = new URL(request.url)

    const posthogUrl = url.pathname.startsWith(`${POSTHOG_API_HOST_PREFIX}/static/`)
        ? new URL(POSTHOG_ASSETS_HREF)
        : new URL(POSTHOG_PUBLIC_HREF)

    const newUrl = new URL(url)
    newUrl.protocol = posthogUrl.protocol
    newUrl.hostname = posthogUrl.hostname
    newUrl.port = posthogUrl.port
    newUrl.pathname = newUrl.pathname.replace(POSTHOG_API_HOST_PREFIX, "")

    const headers = new Headers(request.headers)
    headers.set("host", posthogUrl.hostname)

    const originalHost = request.headers.get("host")
    if (originalHost) {
        headers.set("X-Forwarded-Host", originalHost)
    }

    headers.delete("cookie")
    headers.delete("connection") // https://datatracker.ietf.org/doc/html/rfc7230#section-6.1

    const response = await fetch(newUrl, {
        method: request.method,
        headers,
        ...(!["HEAD", "GET"].includes(request.method)
            ? {
                  body: request.body,
                  duplex: "half",
              }
            : {}),
    })

    const body = await response.text()

    // See https://github.com/nodejs/undici/issues/2514
    const responseHeaders = new Headers(response.headers)
    if (responseHeaders.has("content-encoding")) {
        responseHeaders.delete("content-encoding")
        responseHeaders.delete("content-length")
    }

    return new Response(body, {
        status: response.status,
        statusText: response.statusText,
        headers: responseHeaders,
    })
}

export const Route = createFileRoute("/(api)/api/foundry-bridge/$")({
    server: {
        handlers: {
            ALL: handler,
            GET: handler,
            POST: handler,
            PUT: handler,
            PATCH: handler,
            DELETE: handler,
            OPTIONS: handler,
            HEAD: handler,
        },
    },
})
```

I took the concepts from the node guide on posthog https://posthog.com/docs/advanced/proxy/node

**Comment by hmz22** [comment]
I have better-auth backend and user better-auth client with this rewrite and in serverFn use authClient.getSession and pass cookie and get data with out problem but when try authClient.useSession show like this:
![Screenshot_2025-12-18_100748](https://github.com/user-attachments/assets/c096130a-2d39-42af-96de-eb28edcff071)
and this my setup
```
const config = defineConfig({
  plugins: [
    devtools(),
    nitro({
      routeRules: {
        '/api/**': {
          proxy: 'https://example.com/api/**',
        },
      },
    }),

    // this is the plugin that enables path aliases
    viteTsConfigPaths({
      projects: ['./tsconfig.json'],
    }),
    tailwindcss(),
    tanstackStart(),
    viteReact(),
  ],
})
```
```
export default config
import { createAuthClient } from 'better-auth/react'
import {
  customSessionClient,
  emailOTPClient,
  genericOAuthClient,
  inferAdditionalFields,
  lastLoginMethodClient,
  phoneNumberClient,
  usernameClient,
} from 'better-auth/client/plugins'

export const authClient = createAuthClient({
  baseURL: 'http://localhost:3000',
  basePath: '/api/auth',
  plugins: [
    usernameClient(),
    emailOTPClient(),
    phoneNumberClient(),
    customSessionClient(),
    lastLoginMethodClient(),
    genericOAuthClient(),
  ],
})
```
```
import { Button } from '@/components/ui/button'
import { authClient } from '@/lib/auth'
import { useNavigate } from '@tanstack/react-router'

export default function Logout() {
  const { data } = authClient.useSession()
  console.log('data', data)
  const navigate = useNavigate()
  return (
    <div>
      <Button
        onClick={async () => {
          const { data, error } = await authClient.signOut()
          console.log('logout', data)
          if (data?.success) {
            navigate({
              to: '/login',
            })
          }

          if (error) {
            console.log('error', error)
          }
        }}
      >
        logout
      </Button>
    </div>
  )
}
```
```
const handler = createServerFn().handler(async () => {
  const request= await getRequest()

  const session = await authClient.getSession({
    fetchOptions: {
      headers: {
        cookie: request.headers.get('cookie')?.toString(),
      },
    },
  })

  console.log(session)

  return null
})
```
also, I used @uncvrd setup and get result like image but in next.js with url rewrite everything worked good

**Comment by gluonfield** [comment]
Any updates on this? This is a blocker for me vs NextJS

**Reply by sunxyw** [reply]
I can see there are some good snippets above, e.g. https://github.com/TanStack/router/discussions/2399#discussioncomment-14745607 . Maybe you can take a look and see if they fit your requirement?

**Comment by ahmafi** [comment]
In the latest version (1.132.0) this is working for me:

In `vite.config.ts`:
```typescript
        nitro({
            routeRules: {
                "/api/**": {
                    proxy: "http://localhost:7171/**",
                },
            },
        }),
```

Here is the [docs](https://nitro.build/guide/routing#route-rules) for nitro.

**Reply by doanhtu07** [reply]
This option masks any errors from server into 502 Bad Gateway, so not really good for the client

**Reply by AbbasCiatek** [reply]
yeah +1 502 Bad Gateway Error makes it unusable

**Reply by AbbasCiatek** [reply]
To any one reading this and getting an error 502, the server doesn't mask errors, fetch is throwing an error because of the body being streamed, and it only happens on 401, the solution so far was to add omit credentials in the fetch options.
